File: 민혁/week4/87946_피로도.java
class Solution {

    static int N,count;
    static int[][] D;
    static boolean[] visited;
    static int answer = 0 ;

    public int solution(int k, int[][] dungeons) {


        // 순열 탐색

        N = dungeons.length; // 던전의 수

        D = dungeons; // 던전
        visited = new boolean[N];
        dfs(k,0);

        return answer;
    }
    private static void dfs(int tiredness,int count) {
        // answer = Math.max(answer, count);

        for(int i = 0; i < N; i ++) {
            if(!visited[i] && tiredness >= D[i][0]) {
                visited[i] = true;
                dfs(tiredness - D[i][1], count + 1);
                visited[i] = false;
            }
        }
        answer = Math.max(answer, count);
    }
}

File: 가원/week4/87946_피로도.py
# 프로그래머스 피로도 문제

from typing import List
from itertools import permutations

def solution(k: int, dungeons: List[List[int]]) -> int:
    rows: int = len(dungeons)
    answer_arr: List[int] = []

    for order in permutations(range(rows)):
        count: int = 0
        current_k: int = k
        for i in order:
            if current_k >= dungeons[i][0]:
                current_k -= dungeons[i][1]
                count += 1
            else:
                break

        answer_arr.append(count)

    answer = max(answer_arr)
    return answer


if __name__ == "__main__":
    k: int = 80
    test_list: List[List[int]] = [[80,20],[50,40],[30,10]]
    print(solution(k, test_list))